---
title: "Outline of the R session"
output: 
  html_notebook:
    toc: true
    toc_float: true
    css: style.css
---

```{r include = FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=TRUE, fig.width = 4, 
                      fig.height = 5, warning = FALSE, 
                      message = FALSE, encoding = "UTF-16")
```

## Topics
R topics we will cover in this workshop are lsit below.     

* Outlines  
* Basics of P programming  
  + About R  
  + Installation of R and Rstudio  
  + R objects  
  + Names and values  
  + R object housekeeping  
  + Recommended R programming style  
  + Getting help with R  
  + R package management  
  + R Workspace  
* Data structures  
  + Vectors
  + Lists
  + Factors
  + Matrices and arrays
  + Data frames and the like
  + More on subsetting
* Functions
  + Overview of functions 
  + Create functions
  + Invoking/calling function
  + Lexical scoping
  + Lazy evaluation
  + ... arguments
  + Exiting a function
  + Argument matching rules for function calls
  + Common functions
* Control flow
  + Choices (if-else, switch, ifelse)
  + Loops (for, while, repeat)
* I/O
  + Input
  + Output
  + Viewing data
  + Handling missing values
* Data manipulation
  + Operators
  + Tidy data with tidyr
  + Data manipulation with dplyr
  + Data structure and type conversion
* Data visualization
  + Overview of R graphics systems
  + ggplot2 and the like
* Statistical analysis
  + Univariate statistics
  + Multivariate statistics
* Running R in batch mode
  + Rscript
  + Passing arguments
  + R shell scripting
* Recommendation for further learning
  + Awesome R packages
  + Awesome R books
  + Further learning recommendations


## Install packages to be used in the workshop  
BiocManager::install() can be used to install packages from CRAN and Bioconductor sites.

```bash
# on Linux, where gfortran is used for fortran-based R package installation, run the following code:
module load r/
export f77=gfortran
export FC=gfortran
export FLIBS=-L/usr/local/gfortran/lib
```

```{r}
# Install all required packages for this R session, and later RNA-seq and ChIP-seq data analysis practice.


packages_list <- c("aplpack", "bootstrap", "broom", "ca", "car", "cluster", "colorspace", "ComplexHeatmap", "corrplot", "d3heatmap", "DAAG", "datasets", "dendextend", "devtools", "doBy", "downloader", "dplyr", "factoextra", "FactoMineR", "fpc", "gclus", "ggforce", "ggm", "ggthemes", "gplots", "grDevices", "grid", "gvlma", "hexbin", "Hmisc", "lattice", "leaps", "MASS", "mclust", "methods", "multcomp", "NbClust", "nFactors", "openxlsx", "pastecs", "pheatmap", "polycor", "psych", "pvclust", "Rcmdr", "RColorBrewer", "readxl", "relaimpo", "rgl", "rmarkdown", "scatterplot3d", "sm", "tibble", "tidyverse", "vcd", "vioplot", "XLConnect", "DESeq2", "edgeR", "pca3d", "sva", "limma", "genefilter", "WriteXLS", "DiffBind", "ChIPQC",  "ChIPseeker", "GenomicFeatures", "magick")

if (!require("BiocManager")) {install.packages("BiocManager", dependencies = TRUE)}

for (pkg in packages_list)
{
   if (!pkg %in% installed.packages())
   {
      BiocManager::install(pkg, dependencies = TRUE)
   }
}

# After finishing running the above code, to make sure all packages has been installed, please run the following code snippet. You should see nothing printed out.
for (pkg in packages_list)
{
   if (!pkg %in% installed.packages())
   {
         print(pkg)
    }
}

```

To use devtools, you have to install devtools along with a working development environment.  
* Windows: Install **Rtools**.  
* Mac: Install Xcode from the Mac App Store.  
* Linux: Install a compiler and various development libraries (details vary across differnet flavors of Linux).  


## References  
1. _R in Action, Second Edition: Data analysis and graphics with R_  
By **Robert I. Kabacoff**, May 2015, ISBN 9781617291388.
Website: https://www.manning.com/books/r-in-action-second-edition

2. _Advanced R, First and Second Editions_  
By **Hadley Wickham**
Website: http://adv-r.had.co.nz/, and https://adv-r.hadley.nz/

3. Many R blogs, webpages and tutorials. For time reason, I couldn't appropriately cite them. Sorry about that.


## use R and R packages in the condo cluster
If you have problems in installing those R packages in your own computers, You can use the R and R packages I have installed on the condo HPC during the workshop, though you have to eventually get R and its packages installed in your own computer in the future if you need R for your work.

To use that, please log in your account on <ISU NetID>@condo2017.its.iastate.edu. And use "vi" to open the shell configuration file  ".bash_profile" as follows:

```bash
vi .bash_profile
```
After open the file, hit the "I" key on your keyboard to get into "INSERT" mode and move the cursor to the end of the file and hit the "Enter" key. Then copy the following lines ( CTRL + C for Windows, and Command + C for Mac) and paste (SHIFT + insert for Windows, and Command + V for Mac) to the line where the cursor locates.

```bash
# User specific environment and startup programs
export PATH=$PATH:/bin:/usr/bin/:/usr/local/bin:$HOME/bin/:./:~/.local/bin

### alias to avoid mess up files
alias rm='rm -i'
alias cp='cp -i'
alias mv='mv -i'
alias ll='ls -l'
set -o noclobber

# R library path
R_LIBS=${R_LIBS}:/home/haibol/R/x86_64-pc-linux-gnu-library/3.6

### set colorful prompt
export PS1="\[\e[31;40m\][\[\e[m\]\[\e[32;40m\]\u\[\e[m\]\[\e[31;40m\]@\[\e[m\]\[\e[32;40m\]\h\[\e[m\]\[\e[31;40m\]:\[\e[m\]\[\e[32;40m\]\w\[\e[m\]\[\e[31;40m\]]\[\e[m\]\[\e[31;40m\]\\$\[\e[m\] "
```
Then hit the "ESC", the ":" key and type "wq". The modified file will have been saved.

To make the modification effective, run the following command:

```bash
source .bash_profile
```
No you can load the R and run R interactively, by running the following command.

```bash
#  log in a node for interactive job running
salloc -N 1 -t 05:00:00

# load R3.6.0 module
module load r/3.6.0-py2-fupx2uq

# start R
R
```

## Use the storage space in /freetmp/cktuggle-free
Large temporary space /freetmp is mounted on free nodes and on condodtn (it's not mounted on condo login node). Users should use /freetmp/cktuggle-free. To access that space and use it as working space. Please follow the instruction belows after logging in the head node (login node):

```bash
# you should be at your home directory
pwd 

# replace <your ISU NetID> with your real NetID
mkdir -p /freetmp/cktuggle-free/<your ISU NetID>_work   

# Create a symbolic link to your workspace
ln -s /freetmp/cktuggle-free/<your ISU NetID>_work   ./

# now you can start real work in this directory
cd   <your ISU NetID>_work   
```